mountpoint-s3/src/inode.rs

```diff
@@ -382,77 +370,95 @@ impl SuperblockInner {
             .ok_or(InodeError::InodeDoesNotExist(ino))
     }
 
-    /// Find the inode with the given name in a parent directory and update its `stat`. If the inode
-    /// does not exist, create it if possible.
-    pub fn update_or_insert(
+    /// Update the inode with the given name in a parent directory with the remote data.
+    /// It may update or delete an existing inode, or insert a new one.
+    pub fn update_from_remote(
         &self,
         parent_ino: InodeNo,
         name: &str,
-        stat: InodeStat,
-        kind: InodeKind,
-    ) -> Result<Inode, InodeError> {
-        let parent = {
-            let inodes = self.inodes.read().unwrap();
-            inodes
-                .get(&parent_ino)
-                .cloned()
-                .ok_or(InodeError::InodeDoesNotExist(parent_ino))?
-        };
+        remote: Option<RemoteLookup>,
+    ) -> Result<LookedUp, InodeError> {
+        let parent = self.get(parent_ino)?;
 
         // TODO what if a file was overwritten by a directory on the server side?
         if parent.kind() != InodeKind::Directory {
             return Err(InodeError::NotADirectory(parent_ino));
         }
 
-        // Fast path: try with only a read lock on the directory first
+        // Fast path: try with only a read lock on the directory first.
         {
             let parent_state = parent.inner.sync.read().unwrap();
-            if let Some(inode) = Self::update_if_present(parent_ino, &parent_state, name, kind, &stat)? {
-                return Ok(inode);
+            if let UpdateStatus::Updated(lookedup) = Self::try_update_child(&parent_state, name, &remote)? {
+                return Ok(lookedup);
             }
         }
```

Comment:
We can skip this path if the remote lookup is None since the result will never bee `UpdateStatus::Updated`.

Reply:
Well spotted. But actually we could also have `UpdateStatus::Neither` here and return an error, so probably still worth checking.